controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,292 @@
+	usage, ok := r.SyncMap.Load(storageAutoScaler.Spec.DeviceClass)
+	if !ok {
+		err = fmt.Errorf("osd usage not found for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
+		r.Log.Error(err, "device class not found in sync map")
+		// return nil error as scraper will retry if needed
+		return reconcile.Result{}, nil
+	}
+
+	if usage == nil {
+		err = fmt.Errorf("osd usage is nil for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
+		r.Log.Error(err, "device class not found in sync map")
+		// return nil error as scraper will retry if needed
+		return reconcile.Result{}, nil
+	}
+
+	if checkIfScalingNotRequired(usage, storageAutoScaler.Spec.StorageScalingThresholdPercent) {
+		r.Log.Info("osd usage is less than the threshold", "device class", storageAutoScaler.Spec.DeviceClass, "usage", usage.(float64)*100)
+		return reconcile.Result{}, nil
+	}
+	// scaling is required
+	r.Log.Info("scaling is required", "device class", storageAutoScaler.Spec.DeviceClass)
+
+	// calculate the expectedStorageCapacity
+	startOsdSize, expectedOsdSize, startOsdCount, expectedOsdCount, startStorageCapacity, expectedStorageCapacity := calculateExpectedOsdSizeAndCount(storageCluster, storageAutoScaler)
+
+	if expectedStorageCapacity.Cmp(storageAutoScaler.Spec.StorageCapacityLimit) > 0 {
+		r.Log.Info("storage capacity limit reached")
+		originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+		// update the status
+		bool := true
+		storageAutoScaler.Status.StorageCapacityLimitReached = &bool
+		err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// update the status
+	originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+	storageAutoScaler.Status.Phase = "InProgress"
+	storageAutoScaler.Status.LastExpansion = &ocsv1.LastExpansionStatus{
+		StartOsdCount:           uint16(startOsdCount),
+		ExpectedOsdCount:        uint16(expectedOsdCount),
+		StartOsdSize:            startOsdSize,
+		ExpectedOsdSize:         expectedOsdSize,
+		StartStorageCapacity:    startStorageCapacity,
+		ExpectedStorageCapacity: expectedStorageCapacity,
+		StartTime:               metav1.Now(),
+		CompletionTime:          metav1.Time{},
+	}
+	err = r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// patch the storagecluster with the expected osd size and count
+	originalStorageCluster := storageCluster.DeepCopy()
+	for i, deviceSet := range storageCluster.Spec.StorageDeviceSets {
+		deviceClass := storageAutoScaler.Spec.DeviceClass
+		if deviceClassMatchesDeviceSet(deviceClass, deviceSet) {
+			storageCluster.Spec.StorageDeviceSets[i].DataPVCTemplate.Spec.Resources.Requests["storage"] = expectedOsdSize
+			if startOsdCount != expectedOsdCount {
+				storageCluster.Spec.StorageDeviceSets[i].Count = storageCluster.Spec.StorageDeviceSets[i].Count + 1
```

Comment:
This update of the count of osds can only be hit at most once per horizontal scale and not during status checks, right? If this code path is triggered again during a status check, it would again increase the count.

Reply:
if storageAutoScaler.Status.Phase == "InProgress" || storageAutoScaler.Status.Phase == "Failed" {
		return r.verifyScaling(ctx, storageAutoScaler)
	}
	we will make the phase as `inprogress`,  and at top of reconcile we have the above check, which will take directly to the verify function.
	
	So we are relying on the cr phase  when to proceed and scale and when to just verify